### PR TITLE
Jared/fixed edges v2

### PR DESCRIPTION
**Pull request by jared-m-murphy**
## What type is this change?
- [ ] Bug fix
- [X] New feature

## Checklist before requesting a review
- [X] My code follows the style guidelines of this project
- [X] I have tested on MAC OS system 
- [ ] I have tested on Ubuntu system
- [ ] I have tested on the RIT research cluster and updated the spack packages for compiling
- [X] My changes generate no new warnings
- [ ] If the PR is related to a new feature, I added the running scripts for the new feature

## How Has This Been Tested?
Its initial version was run over 60 times for 10000 max genome and 10 bp iterations.  This version was run on predicting Supp_Fuel_Flow(t+1) for 10000 max genome, number_islands 10 and island_size 10 and bp_iterations 10 and produced the intended result.  All/only GP nodes were used.

### CODE REVIEW

File: multithreaded/examm_mt.cxx [hunk 1]
```diff
@@ -102,7 +102,11 @@ int main(int argc, char** argv) {
     WeightRules* weight_rules = new WeightRules();
     weight_rules->initialize_from_args(arguments);
 
-    RNN_Genome* seed_genome = get_seed_genome(arguments, time_series_sets, weight_rules);
+    string seed_genome_type;
```
Comment (jkarns275) [inline multithreaded/examm_mt.cxx:105]: Does this get used? Might want to remove it otherwise.

File: multithreaded/examm_mt.cxx [hunk 2]
```diff
@@ -102,7 +102,11 @@ int main(int argc, char** argv) {
     WeightRules* weight_rules = new WeightRules();
     weight_rules->initialize_from_args(arguments);
 
-    RNN_Genome* seed_genome = get_seed_genome(arguments, time_series_sets, weight_rules);
```
Comment (jkarns275) [inline multithreaded/examm_mt.cxx:105]: Can probably just revert to this

File: rnn/multiply_node_gp.cxx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        );
+        exit(1);
+    }
+
+    d_bias += (d_input[time] * input_values[time]);
+    for (double& num : ordered_d_input[time]) {
+        num *= d_input[time];
+
+        // most likely gradient got huge, so clip it
+        if (num == NAN || num == -numeric_limits<double>::infinity()) {
```
Comment (jkarns275) [inline rnn/multiply_node_gp.cxx:89]: You can look for nans and infinity values simultaneously with `bool isfinite(double num)` https://en.cppreference.com/w/cpp/numeric/math/isfinite

You can tie this together w/ the sign function to get:
```
if (!isfinite(num))
    num = copysign(1, num) * 1000;
```

File: examm/examm.cxx [hunk 4]
```diff
@@ -921,7 +921,7 @@ RNN_Genome* EXAMM::crossover(RNN_Genome* p1, RNN_Genome* p2) {
     // method
     WeightType weight_initialize = weight_rules->get_weight_initialize_method();
     WeightType weight_inheritance = weight_rules->get_weight_inheritance_method();
-    if (weight_inheritance == weight_initialize) {
+    if ((weight_inheritance == weight_initialize) && (weight_inheritance != WeightType::GP)) {
```
Comment (jkarns275) [inline examm/examm.cxx:924]: Don't need parenthesis

File: rnn/multiply_node_gp.cxx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        for (int j = 0; j < total_inputs; j++) {
+            if (j != i) {
+                total *= ordered_input[time][j];
+            }
+        }
+        ordered_d_input[time].push_back(total * bias);
+    }
+
+#ifdef NAN_CHECKS
+    if (isnan(output_values[time]) || isinf(output_values[time])) {
```
Comment (jkarns275) [inline rnn/multiply_node_gp.cxx:62]: `isfinite`

File: rnn/rnn_edge.cxx [hunk 6]
```diff
@@ -164,12 +165,21 @@ void RNN_Edge::propagate_backward(int32_t time) {
     // Log::trace("propgating backward on edge %d at time %d from node %d to node %d\n", innovation_number, time,
     // output_innovation_number, input_innovation_number);
     double delta;
-    if (output_node->node_type == MULTIPLY_NODE) {
+    if (output_node->node_type == MULTIPLY_NODE || output_node->node_type == MULTIPLY_NODE_GP) {
```
Comment (jkarns275) [inline rnn/rnn_edge.cxx:168]: What is the difference between multiply and multiply GP nodes?

**Review by jkarns275 — APPROVED** [review]
So long as all of the gradient checks pass I think this is good

File: multithreaded/examm_mt.cxx (hunk 1 above)
Comment (jared-m-murphy) [reply multithreaded/examm_mt.cxx:105]: snap, I thought I deleted that good catch

File: multithreaded/examm_mt.cxx (hunk 2 above)
Comment (jared-m-murphy) [reply multithreaded/examm_mt.cxx:105]: yes, just did

File: rnn/multiply_node_gp.cxx (hunk 3 above)
Comment (jared-m-murphy) [reply rnn/multiply_node_gp.cxx:89]: made the change.

File: examm/examm.cxx (hunk 4 above)
Comment (jared-m-murphy) [reply examm/examm.cxx:924]: made the change

File: rnn/multiply_node_gp.cxx (hunk 5 above)
Comment (jared-m-murphy) [reply rnn/multiply_node_gp.cxx:62]: Previous nodes like rnn_node have the form used.  I might just leave it here in this case.

File: rnn/rnn_edge.cxx (hunk 6 above)
Comment (jared-m-murphy) [reply rnn/rnn_edge.cxx:168]: multiply_node has an additive bias and multiply_node_gp has a multiplicative bias.  Also, multiply_node_gp would be flagged for not training its input weights.

File: rnn/multiply_node_gp.cxx (hunk 5 above)
Comment (jkarns275) [reply rnn/multiply_node_gp.cxx:62]: it should probably be changed there too 💀

Comment (jared-m-murphy) [reply rnn/multiply_node_gp.cxx]: alright, I made the change to multiply_node, multiply_node_gp, and rnn_node

File: scripts/node_tests/simple_node_tests.sh
Comment (jkarns275) [inline scripts/node_tests/simple_node_tests.sh:1]: Were the tests you removed here redundant?

**Comment by jkarns275** [comment]
Have you tested vanilla EXAMM after these changes? Including gradient tests and the actual algorithm. Beyond that should be fine.

File: scripts/node_tests/simple_node_tests.sh
Comment (jared-m-murphy) [reply scripts/node_tests/simple_node_tests.sh:1]: they were. there is a sister file in the folder for the gp node gradient tests

**Review by jkarns275 — APPROVED** [review]

**Pull request merged**